pkg/rc/replication_controller.go

```diff
@@ -141,72 +157,129 @@ func (rc *replicationController) meetDesires() error {
+				rc.ReplicasDesired, len(currentNodes), len(eligible), i,
+			)
 		}
-	} else if len(current) > rc.ReplicasDesired {
-		eligible, err := rc.eligibleNodes()
+
+		err := rc.schedule(scheduleOn)
 		if err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		// If we need to downsize the number of nodes, prefer any in current that are not eligible anymore.
-		// TODO: evaluate changes to 'eligible' more frequently
-		preferred := sets.NewString(current...).Difference(sets.NewString(eligible...))
-		rest := sets.NewString(current...).Difference(preferred)
-		toUnschedule := len(current) - rc.ReplicasDesired
-		rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+func (rc *replicationController) removePods(current PodLocations) error {
+	currentNodes := current.Nodes()
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
 
-		for i := 0; i < toUnschedule; i++ {
-			unscheduleFrom, ok := preferred.PopAny()
+	// If we need to downsize the number of nodes, prefer any in current that are not eligible anymore.
+	// TODO: evaluate changes to 'eligible' more frequently
+	preferred := sets.NewString(currentNodes...).Difference(sets.NewString(eligible...))
+	rest := sets.NewString(currentNodes...).Difference(preferred)
+	toUnschedule := len(current) - rc.ReplicasDesired
+	rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+
+	for i := 0; i < toUnschedule; i++ {
+		unscheduleFrom, ok := preferred.PopAny()
+		if !ok {
+			var ok bool
+			unscheduleFrom, ok = rest.PopAny()
 			if !ok {
-				var ok bool
-				unscheduleFrom, ok = rest.PopAny()
-				if !ok {
-					// This should be mathematically impossible unless replicasDesired was negative
-					return util.Errorf(
-						"Unable to unschedule enough nodes to meet replicas desired: %d replicas desired, %d current.",
-						rc.ReplicasDesired, len(current),
-					)
-				}
+				// This should be mathematically impossible unless replicasDesired was negative
+				return util.Errorf(
+					"Unable to unschedule enough nodes to meet replicas desired: %d replicas desired, %d current.",
+					rc.ReplicasDesired, len(current),
+				)
 			}
-			err := rc.unschedule(unscheduleFrom)
+		}
+		err := rc.unschedule(unscheduleFrom)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rc *replicationController) ensureConsistency(current PodLocations) error {
+	manifestSHA, err := rc.Manifest.SHA()
+	if err != nil {
+		return err
+	}
+	for _, pod := range current {
+		intent, _, err := rc.kpStore.Pod(kp.INTENT_TREE, pod.Node, types.PodID(pod.PodID))
+		if err != nil && err != pods.NoCurrentManifest {
+			return err
+		}
+		var intentSHA string
+		if intent != nil {
+			intentSHA, err = intent.SHA()
 			if err != nil {
-				return err
+				rc.logger.WithError(err).WithField("node", pod.Node).Warn("Could not hash manifest to determine consistency of intent")
```

Comment:
Actually, I think what i meant is that `WithError` only behaves properly if its first, since our Logger has special implementation.

Repro case: 

```package main

import (
	"fmt"
	"os"

	"github.com/square/p2/Godeps/_workspace/src/github.com/Sirupsen/logrus"
	"github.com/square/p2/pkg/logging"
	"github.com/square/p2/pkg/util"
)

func main() {
	logger := logging.NewLogger(logrus.Fields{
		"a": 1,
	})
	logger.Logger.Out = os.Stdout

	sublogger := logger.SubLogger(logrus.Fields{
		"b": 2,
	})

	err := util.Errorf("some_error")

	fmt.Println("Using both calls:")
	sublogger.WithField("c", 3).WithError(err).Errorln("using both calls")

	fmt.Println("Using one call:")
	sublogger.WithErrorAndFields(err, logrus.Fields{
		"c": 3,
	},
	).Errorln("using one call")
}
```

output: 
```Using both calls:
{"Counter":0,"PID":33187,"a":1,"b":2,"c":3,"error":"main.go:22: some_error","level":"error","msg":"using both calls","time":"2016-02-22T09:23:59-08:00"}
Using one call:
{"Counter":2,"PID":33187,"a":1,"b":2,"c":3,"err":"main.go:22: some_error","filename":"main.go","function":"main.main","level":"error","line_number":22,"msg":"using one call","time":"2016-02-22T09:23:59-08:00"}
```